File: backend/app/services/compatibility_service.py
from datetime import datetime, timezone

from app.models.chart_models import (
    BirthDetails,
    ClientProfile,
    CompatibilityResponse,
    CompatibilityTheme,
    SynastryAspectInfo,
)
from app.services.chart_service import ChartService

SIGN_ELEMENTS = {
    "Aries": "Fire",
    "Leo": "Fire",
    "Sagittarius": "Fire",
    "Taurus": "Earth",
    "Virgo": "Earth",
    "Capricorn": "Earth",
    "Gemini": "Air",
    "Libra": "Air",
    "Aquarius": "Air",
    "Cancer": "Water",
    "Scorpio": "Water",
    "Pisces": "Water",
}

ASPECT_DEFS = {
    "Conjunction": (0.0, 8.0),
    "Opposition": (180.0, 8.0),
    "Trine": (120.0, 8.0),
    "Square": (90.0, 7.0),
    "Sextile": (60.0, 6.0),
}

HARMONIOUS = {"Conjunction", "Trine", "Sextile"}
PRIORITY_PLANETS = {"Sun", "Moon", "Mercury", "Venus", "Mars", "Jupiter", "Saturn"}


class CompatibilityService:
    def __init__(self) -> None:
        self._chart_service = ChartService()

    @staticmethod
    def _angle_between(lon1: float, lon2: float) -> float:
        diff = abs(lon1 - lon2) % 360.0
        return diff if diff <= 180.0 else 360.0 - diff

    def _build_synastry_aspects(self, primary_planets, partner_planets) -> list[SynastryAspectInfo]:
        aspects: list[SynastryAspectInfo] = []
        for primary in primary_planets:
            if primary.name not in PRIORITY_PLANETS:
                continue
            for partner in partner_planets:
                if partner.name not in PRIORITY_PLANETS:
                    continue
                angle = self._angle_between(primary.longitude, partner.longitude)
                for aspect_type, (target, max_orb) in ASPECT_DEFS.items():
                    orb = abs(angle - target)
                    if orb <= max_orb:
                        aspects.append(
                            SynastryAspectInfo(
                                primary_planet=primary.name,
                                partner_planet=partner.name,
                                aspect_type=aspect_type,
                                orb=round(orb, 4),
                                harmonious=aspect_type in HARMONIOUS,
                            )
                        )
                        break
        aspects.sort(key=lambda item: item.orb)
        return aspects[:18]

    @staticmethod
    def _client_profile_from_birth(birth: BirthDetails) -> ClientProfile:
        return ClientProfile(
            name=birth.name,
            country=birth.country,
            state=birth.state,
            city=birth.city,
            town=birth.town,
            latitude=birth.latitude,
            longitude=birth.longitude,
            timezone_offset_minutes=birth.timezone_offset_minutes,
            time_zone_id=birth.time_zone_id,
        )

    def build_compatibility(self, primary: BirthDetails, partner: BirthDetails) -> CompatibilityResponse:
        primary_chart = self._chart_service.build_chart(
            primary,
            include_transits=False,
            include_premium=False,
            persist_result=False,
            subscription_tier="premium",
        )
        partner_chart = self._chart_service.build_chart(
            partner,
            include_transits=False,
            include_premium=False,
            persist_result=False,
            subscription_tier="premium",
        )

        synastry_aspects = self._build_synastry_aspects(
            primary_chart.chart.planets,
            partner_chart.chart.planets,
        )

        primary_moon = next(planet for planet in primary_chart.chart.planets if planet.name == "Moon")
        partner_moon = next(planet for planet in partner_chart.chart.planets if planet.name == "Moon")
        primary_sun = next(planet for planet in primary_chart.chart.planets if planet.name == "Sun")
        partner_sun = next(planet for planet in partner_chart.chart.planets if planet.name == "Sun")
        primary_venus = next(planet for planet in primary_chart.chart.planets if planet.name == "Venus")
        partner_mars = next(planet for planet in partner_chart.chart.planets if planet.name == "Mars")
        primary_mercury = next(planet for planet in primary_chart.chart.planets if planet.name == "Mercury")
        partner_mercury = next(planet for planet in partner_chart.chart.planets if planet.name == "Mercury")

        harmonious_count = sum(1 for aspect in synastry_aspects if aspect.harmonious)
        tense_count = len(synastry_aspects) - harmonious_count

        score = 58.0
        if SIGN_ELEMENTS[primary_moon.sign] == SIGN_ELEMENTS[partner_moon.sign]:
            score += 12
        if SIGN_ELEMENTS[primary_sun.sign] == SIGN_ELEMENTS[partner_sun.sign]:
            score += 8

        venus_mars_link = next(
            (
                aspect
                for aspect in synastry_aspects
                if aspect.primary_planet == "Venus" and aspect.partner_planet == "Mars"
            ),
            None,
        )
        if venus_mars_link:
            score += 14 if venus_mars_link.harmonious else -8

        moon_links = [
            aspect for aspect in synastry_aspects
            if {aspect.primary_planet, aspect.partner_planet} & {"Moon", "Sun"}
        ]
        if any(aspect.harmonious for aspect in moon_links):
            score += 10

        mercury_link = next(
            (
                aspect
                for aspect in synastry_aspects
                if aspect.primary_planet == "Mercury" and aspect.partner_planet == "Mercury"
            ),
            None,
        )
        if mercury_link:
            score += 8 if mercury_link.harmonious else -5

        score += harmonious_count * 1.5
        score -= tense_count * 1.2
        score = max(20.0, min(98.0, round(score, 1)))

        themes = [
            CompatibilityTheme(
                title="Emotional Rhythm",
                insight=(
                    f"{primary.name}'s Moon in {primary_moon.sign} meets {partner.name}'s Moon in {partner_moon.sign}. "
                    f"Both charts lean into {SIGN_ELEMENTS[primary_moon.sign].lower()} and "
                    f"{SIGN_ELEMENTS[partner_moon.sign].lower()} emotional styles, which shapes how safety and closeness are built."
                ),
                confidence_score=0.86,
            ),
            CompatibilityTheme(
                title="Attraction Pattern",
                insight=(
                    f"{primary.name}'s Venus in {primary_venus.sign} and {partner.name}'s Mars in {partner_mars.sign} "
                    "show the chemistry between affection and desire. Harmonious contact increases natural pull; harder contact creates heat that needs maturity."
                ),
                confidence_score=0.82,
            ),
            CompatibilityTheme(
                title="Communication Style",
                insight=(
                    f"Mercury links between {primary.name} and {partner.name} indicate how quickly misunderstandings get resolved. "
                    f"With Mercury in {primary_mercury.sign} and {partner_mercury.sign}, clarity depends on whether both people process feelings through the same element."
                ),
                confidence_score=0.74,
            ),
        ]

        if tense_count > harmonious_count:
            themes.append(
                CompatibilityTheme(
                    title="Growth Edge",
                    insight="This pairing has real magnetism, but it also carries enough friction that conscious communication and pacing matter. The bond improves when both people build structure around conflict instead of relying on chemistry alone.",
                    confidence_score=0.77,
                )
            )
        else:
            themes.append(
                CompatibilityTheme(
                    title="Long-Term Potential",
                    insight="The synastry leans more supportive than difficult, which usually means this connection has enough ease to grow. Shared rituals, emotional honesty, and clear commitments can turn attraction into durability.",
                    confidence_score=0.79,
                )
            )

        summary = (
            f"{primary.name} and {partner.name} show a compatibility score of {score:.1f}/100. "
            f"The connection is shaped by {harmonious_count} supportive synastry contacts and {tense_count} high-friction contacts."
        )

        return CompatibilityResponse(
            generated_at_utc=datetime.now(timezone.utc),
            primary_client=self._client_profile_from_birth(primary),
            partner_client=self._client_profile_from_birth(partner),
            compatibility_score=score,
            summary=summary,
            themes=themes,
            synastry_aspects=synastry_aspects,
        )

File: lib/engines/compatibility-service.ts

import type { PlanetPosition } from "./swiss-ephemeris-engine";
import { buildChart } from "./chart-service";

// --------------------------------------------------------------------------
// Types
// --------------------------------------------------------------------------

export interface SynastryAspectInfo {
  primary_planet: string;
  partner_planet: string;
  aspect_type: string;
  orb: number;
  harmonious: boolean;
}

export interface CompatibilityTheme {
  title: string;
  insight: string;
  confidence_score: number;
}

export interface ClientProfile {
  name: string;
  country: string;
  state: string;
  city: string;
  town: string;
  latitude: number;
  longitude: number;
  timezone_offset_minutes: number;
  time_zone_id: string;
}

export interface CompatibilityResponse {
  generated_at_utc: string;
  primary_client: ClientProfile;
  partner_client: ClientProfile;
  compatibility_score: number;
  summary: string;
  themes: CompatibilityTheme[];
  synastry_aspects: SynastryAspectInfo[];
  saved_comparison_id: string | null;
}

export interface BirthDetailsInput {
  name: string;
  birth_date: string;
  birth_time: string;
  engine_id?: string;
  timezone_offset_minutes: number;
  latitude: number;
  longitude: number;
  country?: string;
  state?: string;
  city?: string;
  town?: string;
  time_zone_id?: string;
}

// --------------------------------------------------------------------------
// Constants
// --------------------------------------------------------------------------

const SIGN_ELEMENTS: Record<string, string> = {
  Aries: "Fire", Leo: "Fire", Sagittarius: "Fire",
  Taurus: "Earth", Virgo: "Earth", Capricorn: "Earth",
  Gemini: "Air", Libra: "Air", Aquarius: "Air",
  Cancer: "Water", Scorpio: "Water", Pisces: "Water",
};

const ASPECT_DEFS: Record<string, [number, number]> = {
  Conjunction: [0, 8],
  Opposition: [180, 8],
  Trine: [120, 8],
  Square: [90, 7],
  Sextile: [60, 6],
};

const HARMONIOUS = new Set(["Conjunction", "Trine", "Sextile"]);
const PRIORITY_PLANETS = new Set(["Sun", "Moon", "Mercury", "Venus", "Mars", "Jupiter", "Saturn"]);

// --------------------------------------------------------------------------
// Helpers
// --------------------------------------------------------------------------

function angleBetween(lon1: number, lon2: number): number {
  const diff = Math.abs(lon1 - lon2) % 360;
  return diff <= 180 ? diff : 360 - diff;
}

function buildSynastryAspects(
  primaryPlanets: PlanetPosition[],
  partnerPlanets: PlanetPosition[]
): SynastryAspectInfo[] {
  const aspects: SynastryAspectInfo[] = [];

  for (const primary of primaryPlanets) {
    if (!PRIORITY_PLANETS.has(primary.name)) continue;
    for (const partner of partnerPlanets) {
      if (!PRIORITY_PLANETS.has(partner.name)) continue;

      const angle = angleBetween(primary.longitude, partner.longitude);
      for (const [aspectType, [target, maxOrb]] of Object.entries(ASPECT_DEFS)) {
        const orb = Math.abs(angle - target);
        if (orb <= maxOrb) {
          aspects.push({
            primary_planet: primary.name,
            partner_planet: partner.name,
            aspect_type: aspectType,
            orb: Math.round(orb * 10000) / 10000,
            harmonious: HARMONIOUS.has(aspectType),
          });
          break; // Only the first matching aspect per pair
        }
      }
    }
  }

  aspects.sort((a, b) => a.orb - b.orb);
  return aspects.slice(0, 18);
}

function clientProfileFromBirth(birth: BirthDetailsInput): ClientProfile {
  return {
    name: birth.name,
    country: birth.country ?? "",
    state: birth.state ?? "",
    city: birth.city ?? "",
    town: birth.town ?? "",
    latitude: birth.latitude,
    longitude: birth.longitude,
    timezone_offset_minutes: birth.timezone_offset_minutes,
    time_zone_id: birth.time_zone_id ?? "",
  };
}

// --------------------------------------------------------------------------
// Main
// --------------------------------------------------------------------------

export function buildCompatibility(
  primary: BirthDetailsInput,
  partner: BirthDetailsInput
): CompatibilityResponse {
  const primaryChart = buildChart(primary, {
    includeTransits: false,
    includePremium: false,
    includeUltimate: false,
    subscriptionTier: "premium",
  });
  const partnerChart = buildChart(partner, {
    includeTransits: false,
    includePremium: false,
    includeUltimate: false,
    subscriptionTier: "premium",
  });

  const synastryAspects = buildSynastryAspects(
    primaryChart.chart.planets,
    partnerChart.chart.planets
  );

  const primaryMoon = primaryChart.chart.planets.find((p) => p.name === "Moon")!;
  const partnerMoon = partnerChart.chart.planets.find((p) => p.name === "Moon")!;
  const primarySun = primaryChart.chart.planets.find((p) => p.name === "Sun")!;
  const partnerSun = partnerChart.chart.planets.find((p) => p.name === "Sun")!;
  const primaryVenus = primaryChart.chart.planets.find((p) => p.name === "Venus")!;
  const partnerMars = partnerChart.chart.planets.find((p) => p.name === "Mars")!;
  const primaryMercury = primaryChart.chart.planets.find((p) => p.name === "Mercury")!;
  const partnerMercury = partnerChart.chart.planets.find((p) => p.name === "Mercury")!;

  const harmoniousCount = synastryAspects.filter((a) => a.harmonious).length;
  const tenseCount = synastryAspects.length - harmoniousCount;

  let score = 58;
  if (SIGN_ELEMENTS[primaryMoon.sign] === SIGN_ELEMENTS[partnerMoon.sign]) score += 12;
  if (SIGN_ELEMENTS[primarySun.sign] === SIGN_ELEMENTS[partnerSun.sign]) score += 8;

  const venusMarsLink = synastryAspects.find(
    (a) => a.primary_planet === "Venus" && a.partner_planet === "Mars"
  );
  if (venusMarsLink) {
    score += venusMarsLink.harmonious ? 14 : -8;
  }

  const moonLinks = synastryAspects.filter((a) => {
    const pair = new Set([a.primary_planet, a.partner_planet]);
    return pair.has("Moon") || pair.has("Sun");
  });
  if (moonLinks.some((a) => a.harmonious)) score += 10;

  const mercuryLink = synastryAspects.find(
    (a) => a.primary_planet === "Mercury" && a.partner_planet === "Mercury"
  );
  if (mercuryLink) {
    score += mercuryLink.harmonious ? 8 : -5;
  }

  score += harmoniousCount * 1.5;
  score -= tenseCount * 1.2;
  score = Math.max(20, Math.min(98, Math.round(score * 10) / 10));

  const themes: CompatibilityTheme[] = [
    {
      title: "Emotional Rhythm",
      insight: `${primary.name}'s Moon in ${primaryMoon.sign} meets ${partner.name}'s Moon in ${partnerMoon.sign}. Both charts lean into ${SIGN_ELEMENTS[primaryMoon.sign].toLowerCase()} and ${SIGN_ELEMENTS[partnerMoon.sign].toLowerCase()} emotional styles, which shapes how safety and closeness are built.`,
      confidence_score: 0.86,
    },
    {
      title: "Attraction Pattern",
      insight: `${primary.name}'s Venus in ${primaryVenus.sign} and ${partner.name}'s Mars in ${partnerMars.sign} show the chemistry between affection and desire. Harmonious contact increases natural pull; harder contact creates heat that needs maturity.`,
      confidence_score: 0.82,
    },
    {
      title: "Communication Style",
      insight: `Mercury links between ${primary.name} and ${partner.name} indicate how quickly misunderstandings get resolved. With Mercury in ${primaryMercury.sign} and ${partnerMercury.sign}, clarity depends on whether both people process feelings through the same element.`,
      confidence_score: 0.74,
    },
  ];

  if (tenseCount > harmoniousCount) {
    themes.push({
      title: "Growth Edge",
      insight: "This pairing has real magnetism, but it also carries enough friction that conscious communication and pacing matter. The bond improves when both people build structure around conflict instead of relying on chemistry alone.",
      confidence_score: 0.77,
    });
  } else {
    themes.push({
      title: "Long-Term Potential",
      insight: "The synastry leans more supportive than difficult, which usually means this connection has enough ease to grow. Shared rituals, emotional honesty, and clear commitments can turn attraction into durability.",
      confidence_score: 0.79,
    });
  }

  const summary = `${primary.name} and ${partner.name} show a compatibility score of ${score.toFixed(1)}/100. The connection is shaped by ${harmoniousCount} supportive synastry contacts and ${tenseCount} high-friction contacts.`;

  return {
    generated_at_utc: new Date().toISOString(),
    primary_client: clientProfileFromBirth(primary),
    partner_client: clientProfileFromBirth(partner),
    compatibility_score: score,
    summary,
    themes,
    synastry_aspects: synastryAspects,
    saved_comparison_id: null,
  };
}
